### PR TITLE
Lazy data files resolution and offline cache reload

### PR DESCRIPTION
**Pull request by lhoestq**
Includes both https://github.com/huggingface/datasets/pull/6458 and https://github.com/huggingface/datasets/pull/6459

This PR should be merged instead of the two individually, since they are conflicting

## Offline cache reload

 it can reload datasets that were pushed to hub if they exist in the cache.

example:

```python
>>> Dataset.from_dict({"a": [1, 2]}).push_to_hub("lhoestq/tmp")
>>> load_dataset("lhoestq/tmp")
DatasetDict({
    train: Dataset({
        features: ['a'],
        num_rows: 2
    })
})
```

and later, without connection:

```python
>>> load_dataset("lhoestq/tmp")
Using the latest cached version of the dataset since lhoestq/tmp couldn't be found on the Hugging Face Hub
Found the latest cached dataset configuration 'default' at /Users/quentinlhoest/.cache/huggingface/datasets/lhoestq___tmp/default/0.0.0/da0e902a945afeb9 (last modified on Wed Dec 13 14:55:52 2023).
DatasetDict({
    train: Dataset({
        features: ['a'],
        num_rows: 2
    })
})
```

- Updated `CachedDatasetModuleFactory` to look for datasets in the cache at `<namespace>___<dataset_name>/<config_id>`
- Since the metadata configs parameters are not available in offline mode, we don't know which folder to load (config_id and hash change), so I simply load the latest one
  - I instantiate a BuilderConfig even if there is no metadata config with the right config_name
  - Its config_id is equal to the config_name to be able to retrieve it in the cache (no more suffix for configs from metadata configs)
  - We can reload this config if offline mode by specifying the right config_name (same as online !)
- Consequences of this change:
  - Only when there are user's parameters it creates a custom builder config with config_id = config_name + user parameters hash
  - the hash used to name the cache folder takes into account the metadata config and the dataset info, so that the right cache can be reloaded when there is internet connection without redownloading the data or resolving the data files. For local directories I hash the builder configs and dataset info, and for datasets on the hub I use the commit sha as hash.
  - cache directories now look like `config/version/commit_sha` for hub datasets which is clean :)

Fix https://github.com/huggingface/datasets/issues/3547

## Lazy data files resolution

this makes this code run in 2sec instead of >10sec

```python
from datasets import load_dataset

ds = load_dataset("glue", "sst2", streaming=True, trust_remote_code=False)
```

For some datasets with many configs and files it can be up to 100x faster.
This is particularly important now that some datasets will be loaded from the Parquet export instead of the scripts.

The data files are only resolved in the builder `__init__`. To do so I added DataFilesPatternsList and DataFilesPatternsDict that have `.resolve()` to return resolved DataFilesList and DataFilesDict

### CODE REVIEW

**Review by albertvillanova — COMMENTED** [review]
Thanks for the enhancements.

Huge PR to review... I guess you tested all the edge cases.

Naive question: is there any breaking change when loading?

**Comment by lhoestq** [comment]
> Naive question: is there any breaking change when loading?

No breaking changes except that the cache folders are different

e.g. for glue sst2 (has parquet export)

```
This branch
~/.cache/huggingface/datasets/glue/sst2/0.0.0/d60c0e9aaf181d08
On main
~/.cache/huggingface/datasets/glue/sst2/0.0.0/74a75637ac4acd3f
On 2.15.0
~/.cache/huggingface/datasets/glue/sst2/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad
```

e.g. for wikimedia/wikipedia ab (has metadata configs)


```
This branch
~/.cache/huggingface/datasets/wikimedia___wikipedia/20231101.ab/0.0.0/ffe9107c949b550b
On main (takes ages to load)
~/.cache/huggingface/datasets/wikimedia___wikipedia/20231101.ab/0.0.0/cfa627e27933df13
On 2.15.0 (takes ages to load)
~/.cache/huggingface/datasets/wikimedia___wikipedia/20231101.ab/0.0.0/e92ee7a91c466564
```


e.g. for lhoestq/demo1 (no metadata configs)


```
This branch
~/.cache/huggingface/datasets/lhoestq___demo1/default/0.0.0/a426d28fefbc360d
On main (takes ages to load)
~/.cache/huggingface/datasets/lhoestq___demo1/default-8a4a0b7a240d3c5e/0.0.0/eea64c71ca8b46dd3f537ed218fc9bf495d5707789152eb2764f5c78fa66d59d
On 2.15.0 (takes ages to load)
~/.cache/huggingface/datasets/lhoestq___demo1/default-59d4029e0bb36ae0/0.0.0/eea64c71ca8b46dd3f537ed218fc9bf495d5707789152eb2764f5c78fa66d59d
```

File: src/datasets/builder.py [hunk 1]
```diff
@@ -565,6 +570,12 @@ def _create_builder_config(
         if not builder_config.name:
             raise ValueError(f"BuilderConfig must have a name, got {builder_config.name}")
 
+        # resolve data files if needed
+        builder_config._resolve_data_files(
+            base_path=self.base_path,
+            download_config=DownloadConfig(token=self.token, storage_options=self.storage_options),
+        )
```
Comment (lhoestq) [inline src/datasets/builder.py:577]: This is where data files are revolved, this is run in the builder `__init__`

File: src/datasets/load.py [hunk 2]
```diff
@@ -2186,17 +2216,13 @@ def load_dataset_builder(
...
     if (
-        dataset_module.builder_configs_parameters.metadata_configs
-        and config_name in dataset_module.builder_configs_parameters.metadata_configs
+        path in _PACKAGED_DATASETS_MODULES
+        and data_files is None
+        and dataset_module.builder_configs_parameters.builder_configs[0].data_files is None
     ):
-        hash = update_hash_with_config_parameters(
-            hash, dataset_module.builder_configs_parameters.metadata_configs[config_name]
-        )
```
Comment (lhoestq) [inline src/datasets/load.py:2197]: This code is replaced by dedicated `hash` updates in each module factory

File: src/datasets/load.py [hunk 3]
```diff
@@ -1097,6 +1093,7 @@ def get_module(self) -> DatasetModule:
         if default_config_name is None and len(dataset_infos) == 1:
             default_config_name = next(iter(dataset_infos))
 
+        hash = update_hash_for_cache(hash, metadata_configs=metadata_configs, dataset_infos=dataset_infos)
```
Comment (lhoestq) [inline src/datasets/load.py:1096]: here for example

**Comment by lhoestq** [comment]
Arf I just found a bug: if you modify a dataset and reload it from the hub it won't download the new version - I think I need to resolve the data_files earlier in load_dataset_builder and update the hash

File: src/datasets/load.py [hunk 4]
```diff
@@ -2205,7 +2231,17 @@ def load_dataset_builder(
...
     builder_cls = get_dataset_builder_class(dataset_module, dataset_name=dataset_name)
+    if len(builder_cls.builder_configs) > 1:
+        builder_config = builder_cls.builder_configs.get(config_name or builder_cls.DEFAULT_CONFIG_NAME)
+    elif len(builder_cls.builder_configs) == 1:
+        builder_config = builder_cls.BUILDER_CONFIGS[0]
+    else:
+        builder_config = None
+    if builder_config and builder_config.data_files is not None:
+        builder_config._resolve_data_files(base_path=builder_kwargs["base_path"], download_config=download_config)
+        hash = update_hash_for_cache(hash, data_files=builder_config.data_files)
```
Comment (lhoestq) [inline src/datasets/load.py:2244]: The data files are resolved here to get the `hash` in the cache directory tree `config/version/hash`

It also takes into account the metadata configs and the dataset info, and you can find the hash update in the dataset modules factories, like  `hash = update_hash_for_cache(hash, metadata_configs=metadata_configs, dataset_infos=dataset_infos)` in HubDatasetModuleFactoryWithoutScript

**Comment by lhoestq** [comment]
I switched to using the git commit sha for the cache directory, which is now `config/version/commit_sha` :) much cleaner than before.

And for local file it's a hash that takes into account the resolved files (and their last modified dates)

**Comment by lhoestq** [comment]
I also ran the `transformers` CI on this branch and it's green

**Comment by lhoestq** [comment]
FYI `huggingface_hub` will have a release on tuesday/wednesday (will speed up load_dataset data files resolution which is now needed for datasets loaded from parquet export) so we can aim on merging this around the same time and do a release on thursday

File: src/datasets/packaged_modules/cache/cache.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    # get most recent
+    cached_directory_path = Path(sorted(cached_directory_paths, key=_get_modification_time)[-1])
+    version, hash = cached_directory_path.parts[-2:]
+    other_configs = [
+        Path(cached_directory_path).parts[-3]
+        for cached_directory_path in glob.glob(os.path.join(cached_datasets_directory_path_root, "*", version, hash))
+        if os.path.isdir(cached_directory_path)
+    ]
+    if not config_name and len(other_configs) > 1:
+        raise ValueError(
```
Comment (polinaeterna) [inline src/datasets/packaged_modules/cache/cache.py:65]: nit: i think we usually use `textwrap.dedent` for multiline error messages like this for readability

**Review by polinaeterna — COMMENTED** [review]
i love the clearer cache structure! 
i wonder if it's possible to make loading of Hub's datasets backward-compatible (loading from old cache without commit sha if it exists). 
apart from that looks good to me :)

File: src/datasets/packaged_modules/cache/cache.py (hunk 5 above)
Comment (lhoestq) [reply src/datasets/packaged_modules/cache/cache.py:60]: I just checked and we mostly use `\n` actually - not a big deal imo

**Comment by lhoestq** [comment]
Merging this one, and hopefully the cache backward compatibility PR soon too :)

Then it will be release time

**Pull request merged**